helix-lock/src/main/java/org/apache/helix/lock/helix/ZKDistributedNonblockingLock.java

```diff
@@ -31,69 +31,157 @@
+    if (leaseTimeout < 0 || waitingTimeout < 0 || cleanupTimeout < 0) {
+      throw new IllegalArgumentException("Timeout cannot be negative.");
+    }
+    if (priority < 0) {
+      throw new IllegalArgumentException("Priority cannot be negative.");
     }
-    _timeout = timeout;
+    _leaseTimeout = leaseTimeout;
     _lockMsg = lockMsg;
     _userId = userId;
     _baseDataAccessor = baseDataAccessor;
+    _priority = priority;
+    _waitingTimeout = waitingTimeout;
+    _cleanupTimeout = cleanupTimeout;
+    _lockListener = lockListener;
+    _isForceful = isForceful;
   }
 
   @Override
-  public boolean tryLock() {
+  public synchronized boolean tryLock() {
     // Set lock information fields
-    long deadline;
-    // Prevent value overflow
-    if (_timeout > Long.MAX_VALUE - System.currentTimeMillis()) {
-      deadline = Long.MAX_VALUE;
-    } else {
-      deadline = System.currentTimeMillis() + _timeout;
+    _baseDataAccessor.subscribeDataChanges(_lockPath, this);
+    LockUpdater updater = new LockUpdater(
+        new LockInfo(_userId, _lockMsg, getNonOverflowTimestamp(_leaseTimeout), _priority,
+            _waitingTimeout, _cleanupTimeout, null, 0, 0, 0));
+    boolean updateResult = _baseDataAccessor.update(_lockPath, updater, AccessOption.PERSISTENT);
+
+    // Check whether the lock request is still pending. If yes, we will wait for the period
+    // recorded in _pendingTimeout.
+    if (_isPending) {
+      try {
+        wait(_pendingTimeout);
+      } catch (InterruptedException e) {
+        throw new HelixException(
+            String.format("Interruption happened while %s is waiting for the " + "lock", _userId),
+            e);
+      }
+      // We have not acquired the lock yet.
+      if (!_isLocked) {
+        // If the reason for not being able to acquire the lock is due to high priority lock
+        // preemption, directly return false.
+        if (_isPreempted) {
+          _baseDataAccessor.unsubscribeDataChanges(_lockPath, this);
+          return false;
+        } else {
+          // Forceful lock request will grab the lock even the current owner has not finished
+          // cleanup work, while non forceful lock request will get an exception.
+          if (_isForceful) {
+            ZNRecord znRecord = composeNewOwnerRecord();
+            LOG.info("Updating Zookeeper with new owner {} information", _userId);
+            _baseDataAccessor
+                .update(_lockPath, new ZNRecordUpdater(znRecord), AccessOption.PERSISTENT);
+            return true;
+          } else {
+            _baseDataAccessor.unsubscribeDataChanges(_lockPath, this);
+            throw new HelixException("Cleanup has not been finished by lock owner");
+          }
+        }
+      }
     }
-    LockUpdater updater = new LockUpdater(new LockInfo(_userId, _lockMsg, deadline));
-    return _baseDataAccessor.update(_lockPath, updater, AccessOption.PERSISTENT);
+    if (!updateResult) {
```

Comment:
The isLocked is actually inside isPending condition, it checks whether the lock is acquired after the waiting timeout expired. And this updateResult is the immediate response from update operation. So it can be true or false depending on the request's priority and owner. E.g. if the request is not the owner and has a lower priority than the owner, the response will be false, and there is no pending.